### DISCUSSION TITLE
Image.SaveAsPng is slow on Blazor.  Is there an alternative?

### DISCUSSION THREAD

**Discussion by edhalsim** [SixLabors/ImageSharp#1544] [Q&A] [answered]
Hi,
Love your library.  I need to generate at runtime a custom image about 1200x800 that I need to set the "src" property of an "img" tag.  I'm using the GetPixelRowSpan approach to populate all the pixels.  So far so good.  Then, I'm doing:
```
image.SaveAsPng(outStream);
Image64 = "data:image/png;base64,..."
```
The problem is the SaveAsPng is taking several seconds (5-10).  Is there a faster alternative?  Thanks for your help.

**Comment by antonfirsov** [comment]
It shouldn't take that long normally. We need to know more about your environment (os, runtime, hardware etc.)
Further queations typically raised in such cases: How do you benchmark? Are you sure you run in release mode?

**Reply by JimBobSquarePants** [reply]
Yeah, there's absolutely no way it runs this slow in a normal environment. You're talking ~30-40ms normally. 

Also curios why @edhalsim is using their own base 64 string generation. [`ToBase64String(Image, IImageFormat)`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.ImageExtensions.html#SixLabors_ImageSharp_ImageExtensions_ToBase64String_SixLabors_ImageSharp_Image_SixLabors_ImageSharp_Formats_IImageFormat_) exists and appears when using the search in the API docs linked from our readme.

**Reply by edhalsim** [reply]
Here's what I'm doing.  I welcome if there's a better way:
```
            using (MemoryStream outStream = new MemoryStream())
            {
                using (Image<Rgba32> img = new Image<Rgba32>(map.width, map.height))
                {
                    for (int y = 0; y < map.height; y++)
                    {
                        Span<Rgba32> pixelRowSpan = img.GetPixelRowSpan(y);
                        for (int x = 0; x < map.width; x++)
                        {
                            byte alpha = (byte)(map.Palette[map.Background[x, y]] >> 24);
                            byte r = (byte)((map.Palette[map.Background[x, y]] >> 16) & 0xFF);
                            byte g = (byte)((map.Palette[map.Background[x, y]] >> 8) & 0xFF);
                            byte b = (byte)(map.Palette[map.Background[x, y]] & 0xFF);
                            pixelRowSpan[x] = new Rgba32(r, g, b, alpha);
                        }
                    }

                    DateTime startTime = DateTime.Now;
                    img.SaveAsPng(outStream);
                    DateTime endTime = DateTime.Now;
                    var diff = (endTime - startTime).TotalSeconds;
                }
                Image64 = "data:image/bmp;base64," + Convert.ToBase64String(outStream.ToArray());
            }
```
diff is reporting 10 seconds or so.  I'm timing only the SaveAsPng method.  If I use SaveAsBmp then it's under 1 second.  I'm generating the image on a server and transferring it to the web client (Blazor) for display.

**Reply by JimBobSquarePants** [reply]
What's your server specs? Are you running in debug or release? 10 seconds is thousands of times slower than expected.

**Reply by JimBobSquarePants** [reply]
`img.ToBase64String(PngFormat.Instance)` will save you that `ToArray()` call btw.

**Reply by edhalsim** [reply]
Debug vs. Release doesn't seem to matter.  Note that this code is running in the user's browser (Blazor WebAssembly app.), not on the server.  Would it be better to generate the PNG image byte array on the server and transfer that to the client for display?  I was trying to minimize the data travelling across the wire.

**Reply by antonfirsov** [reply]
@edhalsim Blazor is backed by mono, which lacks the runtime features needed to make a fully managed image processing library fast. (eg. SIMD)

You should either process images on the server (with ImageSharp) or use native image processing API-s in the browser.

**Reply by edhalsim** [reply]
I ran a test and you're correct of course.  Yet another challenge I've run into using Blazor.  Thanks for helping.

**Comment by antonfirsov** [comment, marked as answer]
Placing an answer for people hitting this or a similar problem in the future:

We do not recommend using ImageSharp in the browser when performance is important. 

**Update:** After having some [chat on twitter](https://twitter.com/EgorBo/status/1363484736400801796) and doing my own reading, I realized it's not even about SIMD. Client-side Blazor still uses an interpreter, which makes it practically impossible to do any compute-intensive task in the browser. .NET 6 should bring AOT (https://github.com/dotnet/aspnetcore/issues/5466) which will *may* make things better, depending on how generics will be implemented in mono-wasm + AOT.

----
WASM does not support SIMD yet, which makes it impossible to make the library fast on Blazor.
Note that there is a proposal to add SIMD intrinsics to WASM (https://github.com/WebAssembly/simd), but I assume it will take years until it gets runtime support in .NET.
